### ISSUE TITLE
Defaulting to_csv to infer compression

### ISSUE THREAD

**Issue by dhimmel** [pandas-dev/pandas#22004]
This issue follows up on https://github.com/pandas-dev/pandas/pull/17900 by thanks @Dobatymo and @gfyoung with review from @jreback. https://github.com/pandas-dev/pandas/pull/17900 added an `'infer'` option to compression in `_get_handle`. The main user-facing benefit here is that `df.to_csv` will be able to infer compression just like `pandas.read_csv`. However, unlike `read_csv` the default value for compression is `None` rather than `'infer'`

Unfortunately, much of the convenience of `compression='infer'` is lost if you have to explicitly specify it. In summary, I think there is a major convenience to the following command to work and automatically perform gzip compression:

```python
df.to_csv('path.csv.gz')
```

## Compatibility assessment

Defaulting to infer would only affect users who are currently using paths with compression extensions but not actually compressing. That's pretty bad practice IMO. Hence, I'm in favor of breaking backwards compatibility and changing the default for compression to infer. It looks like this would go into the major release 0.24?

**Comment by WillAyd** [comment]
I agree conceptually. Probably need to handle cases where this would potentially conflict with the `compression` argument. PRs welcome

**Comment by dhimmel** [comment]
I am happy to open a PR. I think the solution will be as simple as changing the compression default to infer in:

https://github.com/pandas-dev/pandas/blob/322dbf4f812f47d4eaee5d8afa10e17acf6dbc67/pandas/core/frame.py#L1714-L1716

Looks like `to_pickle` already defaults to infer:

https://github.com/pandas-dev/pandas/blob/322dbf4f812f47d4eaee5d8afa10e17acf6dbc67/pandas/io/pickle.py#L11

`to_json` should also probably be switched to default to infer:

https://github.com/pandas-dev/pandas/blob/322dbf4f812f47d4eaee5d8afa10e17acf6dbc67/pandas/io/json/json.py#L29-L32

I don't think the other [`to_*` methods](https://github.com/pandas-dev/pandas/issues/15008) have a compression argument but I should double check.
